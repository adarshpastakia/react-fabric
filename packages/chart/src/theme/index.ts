/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { DarkThemeBase } from "./darkBase";
import { LightThemeBase } from "./lightBase";
import { ChartPalette } from "./palettes";
import Color from "color";

const makePaletteSequence = (colors: string[]) => {
  return [...colors, ...[...colors.slice(1, -1)].reverse()];
};

export const registerTheme = (
  echarts: AnyObject,
  key: string,
  colors: string[],
) => {
  const base = colors[0];
  const handles = Color(base).mix(Color("#f1f5f9"), 0.15).hex();
  const window = Color(base).alpha(0.1).hexa();
  echarts.registerTheme(`${key}`, {
    ...LightThemeBase,
    dataZoom: {
      ...LightThemeBase.dataZoom,
      fillerColor: window,
      selectedDataBackground: {
        areaStyle: {
          color: base,
        },
        lineStyle: {
          color: Color(base).mix(Color("#0f172a"), 0.35).hex(),
        },
      },
      handleColor: base,
      moveHandleStyle: { color: base },
      emphasis: {
        handleColor: {
          color: handles,
        },
        moveHandleStyle: {
          color: handles,
        },
      },
    },
    color: makePaletteSequence(colors),
  });
  echarts.registerTheme(`${key}_dark`, {
    ...DarkThemeBase,
    dataZoom: {
      ...DarkThemeBase.dataZoom,
      fillerColor: window,
      selectedDataBackground: {
        areaStyle: {
          color: Color(base).mix(Color("#f1f5f9"), 0.35).hex(),
        },
        lineStyle: { color: base },
      },
      handleColor: base,
      moveHandleStyle: { color: base },
      emphasis: {
        handleColor: {
          color: handles,
        },
        moveHandleStyle: {
          color: handles,
        },
      },
    },
    color: makePaletteSequence(colors),
  });
};

export const registerThemes = (echarts: AnyObject) => {
  echarts.registerTheme("default", {
    ...LightThemeBase,
    color: ChartPalette.Default,
  });
  echarts.registerTheme("default_dark", {
    ...DarkThemeBase,
    color: ChartPalette.Default,
  });

  echarts.registerTheme("activity", {
    ...LightThemeBase,
    color: ChartPalette.Activity,
  });
  echarts.registerTheme("activity_dark", {
    ...DarkThemeBase,
    color: ChartPalette.Activity,
  });

  registerTheme(echarts, "qualitative", ChartPalette.Qualitative);
  registerTheme(echarts, "diverging", ChartPalette.Diverging);
  registerTheme(echarts, "sequential", ChartPalette.Sequential);
  registerTheme(echarts, "spectral", ChartPalette.Spectral);
  registerTheme(echarts, "uber", ChartPalette.UberPool);
  registerTheme(echarts, "fireice", ChartPalette.FireIce);
  registerTheme(echarts, "warming", ChartPalette.GlobalWarming);
  registerTheme(echarts, "sunrise", ChartPalette.Sunrise);
  registerTheme(echarts, "ocean", ChartPalette.Ocean);
  registerTheme(echarts, "wine", ChartPalette.Wine);
};
