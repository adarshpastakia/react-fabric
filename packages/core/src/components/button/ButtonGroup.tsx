/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import classNames from "classnames";
import { Children, cloneElement } from "react";
import {
  type AriaProps,
  type ChildrenProp,
  type ColorType,
  type CssProp,
  type RefProp,
  type SizeType,
  type TestProps,
} from "../../types";
import { type Button } from "./Button";
import classes from "./Button.module.css";

export interface ButtonGroupProps
  extends CssProp,
    AriaProps,
    TestProps,
    RefProp<HTMLDivElement>,
    ChildrenProp<React.ReactElement<typeof Button>> {
  /**
   * vertical orientation
   */
  vertical?: boolean;
  /**
   * button styling
   */
  variant?: "outline" | "solid" | "link";
  /**
   * button size
   */
  size?: SizeType;
  /**
   * button color
   */
  color?: ColorType;
  /**
   * full rounded edges
   */
  rounded?: boolean;
  /**
   * disabled state
   */
  disabled?: boolean;
  /**
   * stretch to full width
   */
  fullWidth?: boolean;
}

export const ButtonGroup = ({
  vertical,
  className,
  children,
  variant,
  rounded,
  disabled,
  fullWidth,
  color,
  size,
  ...aria
}: ButtonGroupProps) => {
  return (
    <div
      data-ref="buttonGroup"
      className={classNames(
        classes.buttonGroup,
        className,
        fullWidth ? "flex" : "inline-flex",
        "flex-nowrap align-middle rounded",
        vertical && classes.vertical,
        vertical && "flex-col items-stretch",
      )}
      {...aria}
    >
      {Children.map(children, (child: AnyObject) =>
        cloneElement(child, {
          color,
          size,
          variant,
          disabled,
          rounded,
          fullWidth,
          ...child.props,
        }),
      )}
    </div>
  );
};
