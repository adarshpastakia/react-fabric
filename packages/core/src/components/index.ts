/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export { ActionLabel } from "./actionLabel/ActionLabel";
export {
  AnimationBars,
  AnimationIndicator,
  AnimationSpinner,
  Loading,
  Skeleton,
} from "./animations/Animations";
export { Avatar } from "./avatar/Avatar";
export { AvatarGroup } from "./avatar/AvatarGroup";
export { Badge, getBadgeProps } from "./badge/Badge";
export { Breadcrumbs } from "./breadcrumbs/Breadcrumbs";
export { Button } from "./button/Button";
export { ButtonGroup } from "./button/ButtonGroup";
export { ConfirmButton } from "./button/ConfirmButton";
export { ToggleButtonGroup } from "./button/ToggleButtonGroup";
export { Callout } from "./callout/Callout";
export { Card } from "./card/Card";
export { CardCover } from "./card/CardCover";
export { FlipContent, type FlipContentRef } from "./card/FlipContent";
export { Chip } from "./chip/Chip";
export { Collapsable } from "./collapsable/Collapsable";
export { Divider } from "./divider/Divider";
export { Dropdown, DropdownDismiss } from "./dropdown/Dropdown";
export { DropdownTool } from "./dropdown/DropdownTool";
export { EmptyContent } from "./empty/EmptyContent";
export { Icon } from "./icon/Icon";
export { Image, Video } from "./media/Media";
export { ContextMenu } from "./menu/ContextMenu";
export { Menu } from "./menu/Menu";
export { MenuItem } from "./menu/MenuItem";
export { Meter } from "./meter/Meter";
export { Navigator } from "./navigator/Navigator";
export { Panel } from "./panel/Panel";
export { PanelGroup } from "./panel/PanelGroup";
export { PanelStack } from "./panel/PanelStack";
export { ProgressBar } from "./progress/ProgressBar";
export { ProgressCircle } from "./progress/ProgressCircle";
export { Tab } from "./tabpanel/Tab";
export { TabPanel } from "./tabpanel/TabPanel";
