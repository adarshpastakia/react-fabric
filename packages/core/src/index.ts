/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export * from "./components";
export * from "./core";
export { addTranslationBundle } from "./i18n";
export { Flyout, Modal, Tooltip } from "./overlays";
export * from "./typography";

export * from "./hooks/useClientService";
export * from "./hooks/useDebounce";
export * from "./hooks/useEffectDebugger";
export * from "./hooks/useIsDark";
export * from "./hooks/useIsRtl";
export * from "./hooks/useNavigator";
export * from "./hooks/useOverlayService";
export * from "./hooks/useResize";
export * from "./hooks/useResizeObserver";
export * from "./hooks/useStorage";

export { HotKey } from "./hotkeys/HotKey";
export { HotKeyLabel } from "./hotkeys/HotKeyLabel";

export { CoreIcons } from "./types/icons";

export {
  ApplicationProvider,
  useApplicationContext,
  useGlobals,
  useNotificationService,
} from "./context/Global";
