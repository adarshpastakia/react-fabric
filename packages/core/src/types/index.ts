/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
/* istanbul ignore file */

import { type Placement } from "@floating-ui/react";
import { type JSXElementConstructor, type Ref } from "react";

export type SizeType = "xs" | "sm" | "md" | "lg" | "xl";

export type SizeTypeWithZero = "xs" | "sm" | "md" | "lg" | "xl" | 0;

export type ColorType =
  | "primary"
  | "accent"
  | "info"
  | "danger"
  | "success"
  | "warning";

export type PaletteType =
  | "scarlet"
  | "pumpkin"
  | "marigold"
  | "avacado"
  | "jade"
  | "denim"
  | "iris"
  | "lilac"
  | "coral"
  | "wood"
  | "silver"
  | "steel"
  | "olive"
  | "sand";

export type ColorWithGrays = ColorType | "invert" | "dimmed" | "muted";

export const ThemeColors = [
  "primary",
  "accent",
  "info",
  "danger",
  "success",
  "warning",
  "invert",
  "dimmed",
  "muted",
  "gray",
  "dark",
  "light",
];

export interface TooltipType {
  /**
   * tooltip content
   */
  content?: string | JSX.Element;
  /**
   * tooltip color
   */
  color?: ColorType;
  /**
   * tooltip placement
   */
  placement?: Placement;
}

export interface BadgeType {
  /**
   * badge value
   */
  value?: number | string;
  /**
   * max value for number display
   */
  max?: number;
  /**
   * ping animation
   */
  ping?: boolean;
  /**
   * badge CSS className
   */
  className?: string;
}

export interface CssProp {
  /**
   * css classname(s)
   */
  className?: string;
}

export interface TestProps {
  /**
   * data attribute for test id
   */
  "data-testid"?: string;
  /**
   * data attribute for test value
   */
  "data-test-value"?: AnyObject;
}

export type Elements<T> =
  | React.ReactElement<T>
  | React.ReactPortal
  | React.ReactNode
  | string
  | number
  | boolean
  | null
  | undefined;

export type AriaProps = React.AriaAttributes;

/**
 * For wrapping components
 */

export interface RefProp<T = HTMLElement> {
  ref?: Ref<T>;
}

/* *********** */

// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CallbackReturn<T = unknown> = void | T | Promise<T | void>;

export type ReactTag = keyof JSX.IntrinsicElements | JSXElementConstructor<any>;

export type PropsOf<AsTag extends ReactTag> = AsTag extends React.ElementType
  ? Omit<React.ComponentProps<AsTag>, "ref">
  : never;

export type PolymorphicProps<AsTag extends ReactTag> = Omit<
  PropsOf<AsTag>,
  "children" | "as" | "className"
> & {
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  as?: AsTag;
};

export interface IconProps {
  /**
   * svg path / image url
   */
  icon?: string;
  /**
   * flip icon in rtl
   */
  rtlFlip?: boolean;
  /**
   * icon background color
   */
  iconBg?: ColorType | string;
  /**
   * icon color
   */
  iconColor?: ColorType | string;
}

export interface CollapseProps {
  /**
   * Allow collapsing
   */
  collapsable?: boolean;
  /**
   * Default collapsed
   */
  collapsed?: boolean;
  /**
   * Collapse event
   */
  onCollapse?: (state: boolean) => void;
}
export interface ExpandProps {
  /**
   * Allow expanding
   */
  expandable?: boolean;
  /**
   * Default expanded
   */
  expanded?: boolean;
  /**
   * Expand event
   */
  onExpand?: (state: boolean) => void;
}

export interface TooltipProp {
  /**
   * floating tooltip
   */
  tooltip?: string | TooltipType;
}

export interface ChildProp<T = JSX.Element> {
  children: Elements<T>;
}

export interface ChildrenProp<T = JSX.Element> {
  children: Elements<T> | Array<Elements<T>>;
}

export interface PositionObject {
  top: number;
  left: number;
}

export interface SizeObject {
  width: number;
  height: number;
}
