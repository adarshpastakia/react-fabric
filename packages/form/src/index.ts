/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import "./i18n";

export { Controller } from "./form/Controller";
export { Form, type FormRef } from "./form/Form";

export { useFileUploader } from "./hooks/useFileUploader";
export { useFormBuilder } from "./hooks/useFormBuilder";

export { type FormSchema } from "./types/schema";

export { ArrayInput } from "./input/ArrayInput";
export { Checkbox } from "./input/Checkbox";
export { ColorInput } from "./input/ColorInput";
export { ColorPicker } from "./input/ColorPicker";
export { Field } from "./input/Field";
export { FileInput } from "./input/File";
export { HiddenInput } from "./input/Hidden";
export { Input } from "./input/Input";
export { Number } from "./input/Number";
export { Password } from "./input/Password";
export { Radio } from "./input/Radio";
export { RadioGroup } from "./input/RadioGroup";
export { Range } from "./input/Range";
export { Search } from "./input/Search";
export { Slider } from "./input/Slider";
export { Switch } from "./input/Switch";
export { Textarea } from "./input/Textarea";
export { AutoComplete } from "./select/AutoComplete";
export { DualList } from "./select/DualList";
export { List } from "./select/List";
export { Select } from "./select/Select";
