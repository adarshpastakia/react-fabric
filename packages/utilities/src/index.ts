/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* istanbul ignore file */

export * from "./_ascii";
export * from "./_boundingBox";
export { getBoundingBox, getBox } from "./_boundingBox";
export { default as Countries } from "./_countries";
export type { Country } from "./_countries";
export * from "./_debounce";
export * from "./_dedupe";
export { _fetch as fetch } from "./_fetch";
export { default as FileUtil } from "./_fileType";
export * from "./_format";
export * from "./_getByPath";
export * from "./_hash";
export * from "./_interpolate";
export * from "./_isEqual";
export * from "./_isType";
export * from "./_mergeRefs";
export * from "./_tokenize";
export { getImageColorset } from "./getImageColorset";
export { useLogger } from "./useLogger";
export * from "./yup";

export const EMPTY_ARRAY = [];
Object.freeze(EMPTY_ARRAY);

export const EMPTY_OBJECT = {};
Object.freeze(EMPTY_OBJECT);
