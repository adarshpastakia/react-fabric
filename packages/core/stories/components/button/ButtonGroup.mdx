import {
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Heading,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { Fragment } from "react";
import { ButtonGroup, Collapsable } from "../../../src";
import * as ButtonGroupStories from "./ButtonGroup.stories";
import * as NavigatorStories from "./Navigator.stories";

<Meta of={ButtonGroupStories} />

<Title>Button Group</Title>

<Anchor storyId="ButtonGroup">
  <div className="hidden-anchor">
    <Subheading>Button Group</Subheading>
  </div>
</Anchor>

<Description of={ButtonGroup} />

<Primary />

<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonGroupStories._ButtonGroup} />
</Collapsable>

<Anchor storyId="splitButton">
  <Subheading>Split Button</Subheading>
</Anchor>
<Description of={ButtonGroupStories.SplitButton} />
<Canvas withToolbar of={ButtonGroupStories.SplitButton} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonGroupStories.SplitButton} />
</Collapsable>

<Anchor storyId="stretchButtons">
  <Subheading>Stretched Buttons</Subheading>
</Anchor>
<Description of={ButtonGroupStories.StretchButtons} />
<Canvas withToolbar of={ButtonGroupStories.StretchButtons} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ButtonGroupStories.StretchButtons} />
</Collapsable>

<Anchor storyId="navigator">
  <Subheading>Navigator</Subheading>
</Anchor>
<Description of={NavigatorStories._Navigator} />
<Canvas withToolbar of={NavigatorStories._Navigator} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={NavigatorStories._Navigator} />
</Collapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />

<Anchor storyId="sample">
  <Subheading>Example</Subheading>
</Anchor>

```tsx
import { ButtonGroup } from "@react-fabric/core";

export const Example = () => {
  return (
    <ButtonGroup
      icon="icon svg path"
      altIcon="caret icon"
      color="color"
      variant="style"
      loading={boolean}
      disabled={boolean}
      message="Confirmation message"
      okLabel="Yes"
      cancelLabel="No"
    >
      Button Label
    </Button>
  );
};
```
