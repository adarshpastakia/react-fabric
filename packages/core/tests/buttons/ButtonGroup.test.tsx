/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { composeStories } from "@storybook/react";
import "@testing-library/jest-dom";
import { getByTestId, render } from "@testing-library/react";
import { Button } from "../../src";
import * as stories from "../../stories/components/button/ButtonGroup.stories";

const { Tester } = composeStories(stories);

describe("ButtonGroup", () => {
  it("should render buttons", () => {
    const fragment = render(
      <Tester>
        <Button>First</Button>
        <Button>Second</Button>
      </Tester>,
    );
    expect(
      fragment.container.querySelector("[data-ref='buttonGroup']"),
    ).not.toBeNull();
    expect(
      fragment.container.querySelectorAll(
        "[data-ref='buttonGroup'] > [data-ref='button']",
      ),
    ).toHaveLength(2);
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render full-width", () => {
    const fragment = render(
      <Tester fullWidth data-testid="btngrp">
        <Button>First</Button>
        <Button>Second</Button>
      </Tester>,
    );
    expect(
      fragment.container.querySelector("[data-ref='buttonGroup']"),
    ).not.toBeNull();
    expect(getByTestId(fragment.container, "btngrp")).toHaveClass("flex");
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render vertical", () => {
    const fragment = render(
      <Tester vertical data-testid="btngrp">
        <Button>First</Button>
        <Button>Second</Button>
      </Tester>,
    );
    expect(
      fragment.container.querySelector("[data-ref='buttonGroup']"),
    ).not.toBeNull();
    expect(getByTestId(fragment.container, "btngrp")).toHaveClass("flex-col");
    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });
});
