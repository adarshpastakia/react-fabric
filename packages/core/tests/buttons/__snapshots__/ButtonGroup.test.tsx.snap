// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup should render buttons 1`] = `"<div><div class="flex gap-2 items-center p-4"><div data-ref="buttonGroup" class="buttonGroup inline-flex flex-nowrap align-middle rounded"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">First</label></button></div><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Second</label></button></div></div></div></div>"`;

exports[`ButtonGroup should render full-width 1`] = `"<div></div><div><div class="flex gap-2 items-center p-4"><div data-ref="buttonGroup" class="buttonGroup flex flex-nowrap align-middle rounded" data-testid="btngrp"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 block w-full" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner max-w-full w-full" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">First</label></button></div><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 block w-full" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner max-w-full w-full" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Second</label></button></div></div></div></div>"`;

exports[`ButtonGroup should render vertical 1`] = `"<div></div><div></div><div><div class="flex gap-2 items-center p-4"><div data-ref="buttonGroup" class="buttonGroup inline-flex flex-nowrap align-middle rounded vertical flex-col items-stretch" data-testid="btngrp"><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">First</label></button></div><div class="buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary"><label class="buttonLabel truncate" data-colored-icon="false">Second</label></button></div></div></div></div>"`;
